File: core/src/com/shadyaardvark/geom/Layout.java
package com.shadyaardvark.geom;

import java.util.ArrayList;

public class Layout {
    public final Orientation orientation;
    public final Point size;
    public final Point origin;

    public Layout(Orientation orientation, Point size, Point origin) {
        this.orientation = orientation;
        this.size = size;
        this.origin = origin;
    }

    public Point hexToPixel(Hexagon h) {
        double x = (orientation.f0 * h.q + orientation.f1 * h.r) * size.x;
        double y = (orientation.f2 * h.q + orientation.f3 * h.r) * size.y;
        return new Point((int) x + origin.x, (int) y + origin.y);
    }

    public Point hexCornerOffset(int corner) {
        double angle = 2.0 * Math.PI * (corner + orientation.start_angle) / 6;
        return new Point((int) (size.x * Math.cos(angle)), (int) (size.y * Math.sin(angle)));
    }

    public ArrayList<Point> polygonCorners(Hexagon h) {
        ArrayList<Point> corners = new ArrayList<Point>();
        Point center = hexToPixel(h);
        for (int i = 0; i < 6; i++) {
            Point offset = hexCornerOffset(i);
            corners.add(new Point(center.x + offset.x, center.y + offset.y));
        }
        return corners;
    }
}

File: core/src/com/trashcob/dicewars/math/geom/hexagon/Layout.kt
package com.trashcob.dicewars.math.geom.hexagon

import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Vector2

data class Layout(val orientation: Orientation, val size: Vector2, val origin: Vector2) {
    fun hexagonToPoint(hexagon: Hexagon): Vector2 {
        val x = (orientation.f0 * hexagon.q + orientation.f1 * hexagon.r) * size.x
        val y = (orientation.f2 * hexagon.q + orientation.f3 * hexagon.r) * size.y
        return Vector2(x + origin.x, y + origin.y)
    }

    fun pointToHexagon(point: Vector2): Hexagon {
        val q = orientation.b0 * point.x + orientation.b1 * point.y
        val r = orientation.b2 * point.x + orientation.b3 * point.y
        return closestHexagon(q, r, -q - r)
    }

    fun hexagonCorners(hexagon: Hexagon): List<Vector2> {
        val corners = mutableListOf<Vector2>()
        val center = hexagonToPoint(hexagon)

        (0 until 6)
                .map { hexagonCornerOffset(it) }
                .mapTo(corners) { Vector2(center.x + it.x, center.y + it.y) }

        return corners
    }

    private fun closestHexagon(q: Float, r: Float, s: Float): Hexagon {
        val rq = Math.round(q)
        val rr = Math.round(r)
        val rs = Math.round(s)

        val qDiff = Math.abs(rq - q)
        val rDiff = Math.abs(rr - r)
        val sDiff = Math.abs(rs - s)

        when {
            (qDiff > rDiff) and (qDiff > sDiff) -> return Hexagon(-rr - rs, rr, rs)
            rDiff > sDiff -> return Hexagon(rq, -rq - rs, rs)
            else -> return Hexagon(rq, rs, -rq - rr)
        }
    }

    private fun hexagonCornerOffset(corner: Int): Vector2 {
        val angle = 2f * MathUtils.PI * (orientation.startAngle + corner) / 6f
        return Vector2(size.x * MathUtils.cos(angle), size.y * MathUtils.sin(angle))
    }
}